File: src/main/java/ml/wonwoo/springdatashowcase/product/Product.java
package ml.wonwoo.springdatashowcase.product;

import javax.money.MonetaryAmount;
import javax.persistence.*;
import lombok.Getter;
import lombok.Setter;
import ml.wonwoo.springdatashowcase.support.Identifier;

@MappedSuperclass
public class Product {

    @EmbeddedId
    @AttributeOverride(name = "id", column = @Column(name = "PRODUCT_ID"))
    private Identifier productIdentifier = new Identifier();

    @Getter
    @Setter
    private String name;

    @Getter
    @Setter
    private MonetaryAmount price;

    Product() {}

    public Product(String name, MonetaryAmount price) {
        this.name = name;
        this.price = price;
    }

}

File: kotlin/src/main/kotlin/ml/wonwoo/springdatashowcase/product/Product.kt
package ml.wonwoo.springdatashowcase.product

import ml.wonwoo.springdatashowcase.support.Identifier
import javax.money.MonetaryAmount
import javax.persistence.AttributeOverride
import javax.persistence.Column
import javax.persistence.EmbeddedId
import javax.persistence.MappedSuperclass


@MappedSuperclass
open class Product(

    @EmbeddedId
    @AttributeOverride(name = "id", column = Column(name = "PRODUCT_ID"))
    val productIdentifier: Identifier = Identifier(),

    open val name: String,

    open val price: MonetaryAmount

)

